### PR TITLE
Add test cases for a helper function in controller manager

### PR DESCRIPTION
**Pull request by hainesc**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Add test cases for a helper function `IsControllerEnabled` in controller manager.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by jennybuckley** [comment]
/lgtm
Thanks for adding this!

**Comment by hainesc** [comment]
ping @yue9944882 for approve.

Thanks.

**Comment by hainesc** [comment]
ping @stewart-yu for approve.
Thanks.

**Comment by stewart-yu** [comment]
Thanks! I do not understand the test file you add a little, could you like to paste the coverage output before and after?

**Comment by hainesc** [comment]
> could you like to paste the coverage output before and after?

It does not increase the coverage by code line. `coverage: 23.9% of statements`, both before and after.

The current code enables us run  **All but except** as `IsControllerEnabled("charlie", ["*", "-charlie"], ["some", "else"])` and `IsControllerEnabled("alpha", ["*", "-charlie"], ["some", "else"])`.

The code can be refactored as this if we have no need to run them.
```go
func IsControllerEnabled(name string, disabledByDefaultControllers sets.String, controllers []string) bool {
	for _, ctrl := range controllers {
		if ctrl == name {
			return true
		}
		if ctrl == "-"+name {
			return false
		}
		if ctrl == "*" {
			return !disabledByDefaultControllers.Has(name)
		}
	}
	return false
}
```

File: cmd/controller-manager/app/helper_test.go [hunk 1]
```diff
@@ -60,6 +60,20 @@ func TestIsControllerEnabled(t *testing.T) {
 			disabledByDefaultControllers: []string{"delta", "echo"},
 			expected:                     false,
 		},
+		{
+			name:                         "on by all except",
+			controllerName:               "alpha",
+			controllers:                  []string{"*", "-charlie"},
+			disabledByDefaultControllers: []string{"delta", "echo"},
+			expected:                     true,
+		},
```
Comment (stewart-yu) [inline cmd/controller-manager/app/helper_test.go]: does it diff with the case[on by default]?

File: cmd/controller-manager/app/helper_test.go [hunk 2]
```diff
@@ -60,6 +60,20 @@ func TestIsControllerEnabled(t *testing.T) {
...
+		{
+			name:                         "off by all except",
+			controllerName:               "charlie",
+			controllers:                  []string{"*", "-charlie"},
+			disabledByDefaultControllers: []string{"delta", "echo"},
```
Comment (stewart-yu) [inline cmd/controller-manager/app/helper_test.go]: does it diff with the case[off by name]?

**Review by stewart-yu — COMMENTED** [review]
also confused about the test case

File: cmd/controller-manager/app/helper_test.go (hunk 2 above)
Comment (hainesc) [reply cmd/controller-manager/app/helper_test.go]: Yes, all except {"*", "-charlie"} is different from {"alpha", "bravo" "-charlie"}, both in semantical and logical. The var `hasStar` in the function is different, there are `true` and `false` respectively

File: cmd/controller-manager/app/helper_test.go (hunk 1 above)
Comment (hainesc) [reply cmd/controller-manager/app/helper_test.go]: `on by all except` tests `all but not disabled by sign(-)` in `controllers`.

`on by default` tests all but not in `disabledByDefaultControllers`

They are different by where disable, one in `controllers`, one in `disabledByDefaultControllers`

**Comment by hainesc** [comment]
Given: 
a -> all
b -> disabled in parameter `controllers`, that is "-something"
c -> disableByDefaultControlers, this is "something" in parameter `disabledByDefaultControllers `

We should test: 
1. a && !b && c
2. a && b && !c

**Comment by stewart-yu** [comment]
> The code can be refactored as this if we have no need to run them.
> 
> ```go
> func IsControllerEnabled(name string, disabledByDefaultControllers sets.String, controllers []string) bool {
> 	for _, ctrl := range controllers {
> 		if ctrl == name {
> 			return true
> 		}
> 		if ctrl == "-"+name {
> 			return false
> 		}
> 		if ctrl == "*" {
> 			return !disabledByDefaultControllers.Has(name)
> 		}
> 	}
> 	return false
> }
> ```

This refactor is wrong when `controller` contains `*`
Given:
controllers := []string{"*", "-a", "-b"}
name := a
disabledByDefaultControllers := []string["e", "f", "g"]

actually, `-a` and `-b` in controllers is unused,  function `IsControllerEnabled ` return `true` yet? but we want `false`

**Comment by hainesc** [comment]
@stewart-yu 

Yes, the refactor one is wrong, but the current unit tests can not tell that. The refactor function can pass all current test cases.
This is the reason why I add this two test cases.
By adding this two test cases, the refactor function can not pass the test case.

File: cmd/controller-manager/app/helper_test.go (hunk 2 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/helper_test.go]: the key is `-Charlie` in controllers

**Comment by stewart-yu** [comment]
leave other guys for review

**Comment by hainesc** [comment]
ping @deads2k for approve.

Thanks!

**Comment by hainesc** [comment]
ping @deads2k @cheftako @sttts for approve.

File: cmd/controller-manager/app/helper_test.go [hunk 3]
```diff
@@ -60,6 +60,20 @@ func TestIsControllerEnabled(t *testing.T) {
...
 		},
+		{
+			name:                         "on by all except",
+			controllerName:               "alpha",
+			controllers:                  []string{"*", "-charlie"},
+			disabledByDefaultControllers: []string{"delta", "echo"},
+			expected:                     true,
+		},
+		{
+			name:                         "off by all except",
```
Comment (cheftako) [inline cmd/controller-manager/app/helper_test.go]: name seems confusing for the case.
The case seems to be you have set it default on and then explicitly turned off the relevant controller.
How about something like "on by default, off by name"?
I like the test case, it makes the precedence order very clear.

Comment (hainesc) [reply cmd/controller-manager/app/helper_test.go]: Done. I changed to `on by star, not off by name` and `off by name with star`

**Review by cheftako — COMMENTED** [review]
/lgtm

**Comment by hainesc** [comment]
@cheftako thanks for your `lgtm`. 
May you pls consider a `approved` label as other approvers are busy, 
Or call somebody else take a look?

**Pull request merged**